File: typescript/src/main/vsdk/toolkit/io/context/VertexContext.ts
import { Vector3Dd } from "../../common/linealAlgebra/Vector3Dd";
import { Vertex } from "../../skin/Vertex";

export class VertexContext {
  public p: Vector3Dd;
  public n: Vector3Dd;
  public xid: number;
  public clock: number;
  public vertex: Vertex | null;

  public constructor();
  public constructor(inP: Vector3Dd, inN: Vector3Dd, inXid: number, inClock: number, inVertex: Vertex | null);
  public constructor(inP?: Vector3Dd, inN?: Vector3Dd, inXid?: number, inClock?: number, inVertex?: Vertex | null) {
    this.p = new Vector3Dd();
    this.n = new Vector3Dd();
    this.xid = 0;
    this.clock = 0;
    this.vertex = null;

    if (inP !== undefined && inN !== undefined && inXid !== undefined && inClock !== undefined) {
      this.p.copy(inP);
      this.n.copy(inN);
      this.xid = inXid;
      this.clock = inClock;
      this.vertex = inVertex ?? null;
    }
  }

  public copy(source: VertexContext | null): void {
    if (source === null) {
      return;
    }
    this.p.copy(source.p);
    this.n.copy(source.n);
    this.xid = source.xid;
    this.clock = source.clock;
    this.vertex = source.vertex;
  }
}

File: java/base/src/main/java/vsdk/toolkit/io/context/VertexContext.java
package vsdk.toolkit.io.context;

import vsdk.toolkit.common.linealAlgebra.Vector3Dd;
import vsdk.toolkit.environment.geometry.elements.Vertex;

public class VertexContext {
    public Vector3Dd p; // Point
    public Vector3Dd n; // Normal
    public long xid; // Transform id of transform last time the vertex was modified (or created)
    public int clock; // Incremented each change -- resettable
    public Vertex vertex;

    public VertexContext() {
        p = new Vector3Dd();
        n = new Vector3Dd();
        xid = 0;
        clock = 0;
        vertex = null;
    }

    public VertexContext(Vector3Dd inP, Vector3Dd inN, long inXid, int inClock, Vertex inVertex) {
        p = new Vector3Dd();
        n = new Vector3Dd();
        xid = 0;
        clock = 0;
        vertex = null;
        p.copy(inP);
        n.copy(inN);
        xid = inXid;
        clock = inClock;
        vertex = inVertex;
    }

    public void copy(VertexContext source) {
        if (source == null) {
            return;
        }
        p.copy(source.p);
        n.copy(source.n);
        xid = source.xid;
        clock = source.clock;
        vertex = source.vertex;
    }
}
